models/gif_psc_exp.cpp

```diff
@@ -0,0 +1,486 @@
+}
+
+void
+nest::gif_psc_exp::State_::get( DictionaryDatum& d, const Parameters_& p ) const
+{
+  def< double >( d, names::V_m, y3_ );    // Membrane potential
+  def< double >( d, names::E_sfa, sfa_ ); // Adaptive threshold potential
+  def< double >( d, names::stc, stc_ );   // Spike triggered current
+}
+
+void
+nest::gif_psc_exp::State_::set( const DictionaryDatum& d, const Parameters_& p )
+{
+  updateValue< double >( d, names::V_m, y3_ );
+  sfa_stc_initialized_ =
+    false; // vectors of the state should be initialized with new parameter set.
+}
+
+nest::gif_psc_exp::Buffers_::Buffers_( gif_psc_exp& n )
+  : logger_( n )
+{
+}
+
+nest::gif_psc_exp::Buffers_::Buffers_( const Buffers_&, gif_psc_exp& n )
+  : logger_( n )
+{
+}
+
+/* ----------------------------------------------------------------
+ * Default and copy constructor for node
+ * ---------------------------------------------------------------- */
+
+nest::gif_psc_exp::gif_psc_exp()
+  : Archiving_Node()
+  , P_()
+  , S_()
+  , B_( *this )
+{
+  recordablesMap_.create();
+}
+
+nest::gif_psc_exp::gif_psc_exp( const gif_psc_exp& n )
+  : Archiving_Node( n )
+  , P_( n.P_ )
+  , S_( n.S_ )
+  , B_( n.B_, *this )
+{
+}
+
+/* ----------------------------------------------------------------
+ * Node initialization functions
+ * ---------------------------------------------------------------- */
+
+void
+nest::gif_psc_exp::init_state_( const Node& proto )
+{
+  const gif_psc_exp& pr = downcast< gif_psc_exp >( proto );
+  S_ = pr.S_;
+}
+
+void
+nest::gif_psc_exp::init_buffers_()
+{
+  B_.spikes_ex_.clear(); // includes resize
+  B_.spikes_in_.clear(); // includes resize
+  B_.currents_.clear();  //!< includes resize
+  B_.logger_.reset();    //!< includes resize
+  Archiving_Node::clear_history();
+}
+
+void
+nest::gif_psc_exp::calibrate()
+{
+
+  B_.logger_.init();
+
+  double_t h = Time::get_resolution().get_ms();
+  V_.rng_ = kernel().rng_manager.get_rng( get_thread() );
+
+  V_.P11ex_ = std::exp( -h / P_.tau_ex_ );
+  V_.P11in_ = std::exp( -h / P_.tau_in_ );
+
+  double_t tau_m = P_.c_m_ / P_.g_L_;
+
+  // these are determined according to a numeric stability criterion
+  V_.P21ex_ = propagator_32( P_.tau_ex_, tau_m, P_.c_m_, h );
+  V_.P21in_ = propagator_32( P_.tau_in_, tau_m, P_.c_m_, h );
+
+
+  V_.P33_ = std::exp( -h / tau_m );
+  V_.P30_ = 1 / P_.c_m_ * ( 1 - V_.P33_ ) * tau_m;
+  V_.P31_ = ( 1 - V_.P33_ );
+
+  V_.RefractoryCounts_ = Time( Time::ms( P_.t_ref_ ) ).get_steps();
+  assert( V_.RefractoryCounts_
+    >= 0 ); // since t_ref_ >= 0, this can only fail in error
+
+  // initializing internal state
+  if ( !S_.sfa_stc_initialized_ )
+  {
+    V_.P_sfa_.clear();
+    V_.P_stc_.clear();
+    S_.sfa_elems_.clear();
+    S_.stc_elems_.clear();
+
+    for ( size_t i = 0; i < P_.tau_sfa_.size(); i++ )
+    {
+      V_.P_sfa_.push_back( std::exp( -h / P_.tau_sfa_[ i ] ) );
+      S_.sfa_elems_.push_back( 0.0 );
+    }
```

Comment:
As a design principle, `calibrate()` shall only change members of `V_`, never members of the observable state `S_`. I do not think it would be an unreasonable restriction that the *number* of elements of stc and sfa are fixed once simulation of a neuron has started. The `S_.sfa_elems_` and `S.stc_elems_` vectors should therefore be initialized in `init_state_()`. I don't think the `sfa_stc_initialized_`member is required: once the `*_elems_` members are non-empty, they are initialized. Initialization can be done faster using `S_.sfa_elems_.resize( <new size>, 0.0 );`.

On the other hand, the `V_.P_sfa_` and `V_.P_sfc_` vectors must be re-computed on *every* call to `calibrate()`,  since the use may have change the values of the time constants in between `Simulate` calls. We don't need a `clear()`, the following should do:
```c++
V_.P_sfa_.resize( P_.tau_sfa_.size(), 0.0 );
for ( size_t i = 0; i < P_.tau_sfa_.size(); i++ )
{
  V_.P_sfa_[i] = std::exp( -h / P_.tau_sfa_[ i ] );
}
```